### ISSUE TITLE
How to train on own dataset ?

### ISSUE THREAD

**Issue by andohuman** [ElementAI/LCFCN#3]
I wanna try training LCFCN on my own dataset. What are the things I should be looking at (images, annotations, etc) to train the model on my own dataset ?

**Comment by IssamLaradji** [comment]
The `__getitem__` function in the dataset loaders such as in `datasets/trancos.py` shows you what `LCFCN` and its loss expect.  They expect the following items:
```
 return {"images":image, "points":points, 
                "counts":counts, "index":index,
                "image_path":self.path + name + ".jpg"}
```
where images is (1,3,H,W), points is (1,H,W), counts is (1, K), index is the `image id`; also,

H: is the image height
W: is the image width
K: is the number of classes

**Comment by IssamLaradji** [comment]
Happy to help!

> When you say a single point for each object... does that mean something like the center point of each object ? 

Yes, you can take the center of the object as a single point, just like the example you showed. The value of the point represents the class of the object.

**Comment by andohuman** [comment]
Okay I've seen the trancos.py file and mostly understand what's happening. 

1) What are the .mat files that are being loaded ?
2) Also what are the .txt files present in the images directory ? (I'm looking at the TRANCOS dataset)
2) What exactly is the transform function ?

**Comment by IssamLaradji** [comment]
1. the `.mat` files are binary matrices that represent the regions of interest in the image.  Not all datasets have that, for example, `shanghai.py` doesn't have that.

2. the `.txt` files contain the paths to the images which you use to load image at  every iteration by calling `__getitem__`

3. the transform function are used to flip, rotate, or/and normalize the image. `normalization` is important if you are using a pretrained network like `resnet` which expects a specific kind of input distribution.

**Comment by andohuman** [comment]
1) So I don't necessarily need to have regions of interests included in my training images right ?

2) I get the .txt files in image_sets/*.txt, but the .txt files in images/ which have the same name as that of the image. They have some numbers in them.

**Comment by IssamLaradji** [comment]
1. You are right, you don't need regions of interests for your training images;
2. The image_sets specify which image files are for training, validation, and testing. So for training, you only load the images mentioned in `image_sets'train.txt` from `images/`

**Comment by andohuman** [comment]
Man, the last few days I've been breaking my head over this. I don't exactly "get" the loss function (All 4 losses) or how you implemented it in torch. I was hoping if I get the loss function I could write it in keras (which I'm comfortable with). Is there maybe another source (like a blog post or an article) that explains how you practically implemented the loss (and the entire model in general) ?

Thanks a ton for helping out!

**Comment by IssamLaradji** [comment]
you are welcome! you are free to open another issue where i can explain each part of the loss and/or architecture for you.

I don't think there is another source yet, but I am planning to create a blog post on this at some point. Sorry :(

**Comment by tongpinmo** [comment]
> 1. You are right, you don't need regions of interests for your training images;
> 2. The image_sets specify which image files are for training, validation, and testing. So for training, you only load the images mentioned in `image_sets/train.txt` from `images/`
　
I have images and the points files ,should I produce the dots.png and .mat files ?

**Comment by tongpinmo** [comment]
> 1. the `.mat` files are binary matrices that represent the regions of interest in the image.  Not all datasets have that, for example, `shanghai.py` doesn't have that.
> 2. the `.txt` files contain the paths to the images which you use to load image at  every iteration by calling `__getitem__`
> 3. the transform function are used to flip, rotate, or/and normalize the image. `normalization` is important if you are using a pretrained network like `resnet` which expects a specific kind of input distribution.

Actually ,in shanghai.py , line 45, there are .mat files?So ,what's the meaning of `shanghai.py` doesn't have that as you mentioned

**Comment by GloriaZLQ** [comment]
how can i get the file like"pointDict.json" in "  VOCtrainval_11-May-2012\VOCdevkit\VOC2012"?
I'm making my own dataset for training.
